### PR TITLE
Streamline KindOf flags of hulk objects

### PR DESCRIPTION
**Pull request by xezon**
**Merge with Rebase**

This change attempts to fix some hulk flags misconfigurations.

* Hulks of USA Microwave, Humvee, China Troop Crawler(s), Listening Outpost have obsolete IMMOBILE flag removed
* All hulks can now cast reflections

### CODE REVIEW

**Comment by commy2** [comment]
UNATTACKABLE does nothing if no ActiveBody (or similar) module exists.

**Comment by commy2** [comment]
What does "casting reflections" mean? I don't remember reflections in the game.

**Comment by xezon** [comment]
> What does "casting reflections" mean? I don't remember reflections in the game.

`CAN_CAST_REFLECTIONS` means that an object can draw in mirrors. I do not know about such mirrors, but I see corresponding code in Thyme.

**Comment by commy2** [comment]
I wouldn't change UNATTACKABLE and CAN_CAST_REFLECTIONS flags if these don't do anything.

**Comment by xezon** [comment]
I will leave `CAN_CAST_REFLECTIONS`. If the vehicles cast reflections, then logically so should the hulks.

**Comment by xezon** [comment]
I will revert the `UNATTACKABLE` if not needed.

**Comment by xezon** [comment]
Added `UNATTACKABLE` flags are removed.

**Comment by Stubbjax** [comment]
> All vehicle hulks can now bounce

What does this look like?

(This line is also a bit unclear, as the same can be said whether you changed one hulk or all of them.)

**Comment by xezon** [comment]
The bounce change was affecting USA Microwave, Humvee, China Troop Crawler(s), Listening Outpost only.

I took another look at bouncing and was not able to observe difference with removed `IMMOBILE` flag, neither with `AllowBouncing` set to No.

So looks like it does nothing tangible. I renamed the commit accordingly.

## Original

https://user-images.githubusercontent.com/4720891/216760009-0243a08d-5ceb-4702-ba32-5685d1c45ef8.mp4

## Patched, Immobile flag removed

https://user-images.githubusercontent.com/4720891/216760024-9f7bec8b-4b76-48ed-ae26-cf8aa61a2d34.mp4

**Comment by Stubbjax** [comment]
Hmm, there doesn't seem to be a lot of difference. The bounces might look a bit smoother on the second video, but it's hard to tell whether that's just coincidence. If there is no tangible benefit, maybe it is worth considering reapplying `IMMOBILE` to all hulks for performance reasons (assuming this flag negates some physics calculations)?

**Comment by xezon** [comment]
The vast majority of hulks do not have IMMOBILE flag. I would need to get better understanding of IMMOBILE to judge whether or not hulks should have it.

The relevant code is in `PhysicsBehavior::On_Collide` function. It is not yet clear to me that code is 100% correct, because `immobile1` variable is unused.

https://github.com/TheAssemblyArmada/Thyme/blob/develop/src/game/logic/object/update/physicsupdate.cpp#L447-L576

**Comment by Stubbjax** [comment]
From testing, it doesn't seem to have any effect on physics.

In the below footage, the bottom Overlord has `IMMOBILE` while the top Overlord does not. They both bounce around in the same manner.

https://user-images.githubusercontent.com/11547761/216765174-248bfa18-3c64-4697-8270-e0207940d295.mp4

**Comment by xezon** [comment]
I tested Humvee and Ambulance hulks and put breakpoint in `PhysicsBehavior::On_Collide` function. Behaviour is exact same, because `AIUpdateInterface* update` will be null, and `PhysicsBehaviour* physics` will be not null. In this scenario, IMMOBILE flag has no relevance in code path.

Therefore IMMOBILE flag is obsolete and can be removed.

debug

File: Patch104pZH/GameFilesEdited/Data/INI/Object/Hulk.ini [hunk 1]
```diff
@@ -19,7 +19,9 @@ Object DeadMicrowaveHulk
   ; *** AUDIO Parameters ***
   ; *** ENGINEERING Parameters ***
   RadarPriority = UNIT
-  KindOf = CAN_CAST_REFLECTIONS IMMOBILE NO_COLLIDE HULK UNATTACKABLE
+
+  ; Patch104p @bugfix xezon 28/01/2023 Remove immobile flag to allow bouncing.
```
Comment (Stubbjax) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/Hulk.ini]: These comments might not be so accurate. All `IMMOBILE` seems to do is prevent move orders. The `AllowBouncing = Yes` in the `PhysicsBehavior` seems likely to be responsible for the bouncing.

**Comment by xezon** [comment]
I need to merge this soon because I am already working on next fixes and piling changes on unit hulks will make testing difficult.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/Hulk.ini (hunk 1 above)
Comment (xezon) [reply Patch104pZH/GameFilesEdited/Data/INI/Object/Hulk.ini]: Fixed the comments.

**Comment by xezon** [comment]
I have now isolated the NO_COLLIDE flags fix for Militia Tank from this Pull, because that is actually a tangible bug that should have its own Pull.

**Comment by xezon** [comment]
This change looks good to me. Helicopters still have IMMOBILE flag. Maybe can remove too in another change.

**Comment by Stubbjax** [comment]
> I need to merge this soon because I am already working on next fixes and piling changes on unit hulks will make testing difficult.

This is a poor reason to bypass the review process.

> Fixed the comments.

This branch was not approved for merge. You gave me just six minutes to approve after addressing my feedback. You also rewrote git history which is extremely dangerous and terrible practice. How am I supposed to review the change in relation to the review?

**Pull request merged**